### DISCUSSION TITLE
Automated Installer .msi creation compatible with Toolset v4 / HeatWave ?

### DISCUSSION THREAD

**Discussion by drops93** [wixtoolset/issues#7732] [Questions]
I tried to configure my .wixproj file so that it will automatically go through my project files and create a component group which I can then import into my Package.wxs.

Is this not compatible, or what is wrong with my configuration?
I also tried to follow all the debugging suggestions that ChatGPT offers me, but there is no successful building...

My .wixproj files looks like this:
```xml
<Project Sdk="WixToolset.Sdk/4.0.2">
  <PropertyGroup>
    <!-- Define the path to heat.exe -->
    <HeatToolPath Condition="'$(HeatToolPath)' == ''">$(WixToolPath)bin\heat.exe</HeatToolPath>
	  
    <!-- Define the source directory and output file -->
    <SourceConnections>$(SolutionDir)\Connections</SourceConnections>
    <OutputConnectionsFile>$(ProjectDir)\Connections_files.wxs</OutputConnectionsFile>
    <SourceScripts>$(SolutionDir)\Scripts</SourceScripts>
    <OutputScriptsFile>$(ProjectDir)\Scripts_files.wxs</OutputScriptsFile>
  </PropertyGroup>

  <Target Name="BeforeBuild" BeforeTargets="Build">
    <!-- Run heat.exe as a pre-build event for connections -->
    <Exec Command="echo Running heat.exe for Connections" />
    <Exec Command="$(HeatToolPath) dir $(SourceConnections) -gg -g1 -scom -sfrag -srd -cg Connections_files -out $(OutputConnectionsFile)" />
    <!-- Run heat.exe as a pre-build event for scripts -->
    <Exec Command="echo Running heat.exe for Scripts" />
    <Exec Command="$(HeatToolPath) dir $(SourceScripts) -gg -g1 -scom -sfrag -srd -cg Scripts_files -out $(OutputScriptsFile)" />
  </Target>
</Project>
```

And the Package.wxs looks like that:
```xml
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
	<Package Name="$(ProjectName)" Manufacturer= "Testcompany" Version="1.0.0.0" UpgradeCode="is not empty">
		<MajorUpgrade AllowDowngrades="no" DowngradeErrorMessage="!(loc.DowngradeError)" AllowSameVersionUpgrades="no" />
		<MediaTemplate EmbedCab="yes" />


		<Feature Id="Main">
			<!-- Define a Component for the single file -->
			<Component Directory="ConfigFolder">
				<File Source="../TestConfig.xml" />
			</Component>

			<ComponentGroupRef Id="Connections_files" />
			<ComponentGroupRef Id="Scripts_files" />
		</Feature>
	</Package>
</Wix>
```

The folder configuration looks like that: 
```xml
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
	<Fragment>
		<StandardDirectory Id="ProgramFiles6432Folder">
			<Directory Id="INSTALLFOLDER" Name="TestApp">
				<Directory Id="ConfigFolder" Name="TestConfig">
					<Directory Id="Connections" Name="Connections" />
					<Directory Id="Scripts" Name="Scripts" />
				</Directory>
			</Directory>
		</StandardDirectory>
	</Fragment>
</Wix>
```

**Comment by chrpai** [comment]
With certain exceptions, I'm generally not a fan of dynamically authoring an installer project this way. There are just too many things that can go wrong.  I suggest looking at my OSS tool  IsWiX with tutorials here at https://www.github.com/iswix-llc/iswix-tutorials.

I promise that you'll find this a lot easier then trying to ask an AI how to develop WiX.  Hit me up if you'd like a demo or have any questions after.

**Reply by drops93** [reply]
I agree that multiple thing could happened that the installer miss some files or something like that. Maybe I explain a bit more about my application, to understand what I'm looking at. 

I have a generic main software application and then a configurations software, which allows me to use the generic main software in multiple ways. For the main software I use a similiar installer addon. But for the configurations software I search for a way to collect all .dll 's and all content files from my project. 

Why? --> Because the way the configurations software will be programmed is basically the same, only sometimes I have more or less files which the installers needs to copy to the main application locations. So this will then increase the setup time of an configurations project.

Maybe you have an easier way to do this or what is your experience in this ?

**Comment by vdschuck** [comment]
Hey @drops93, 

In Wix v4 heat is a Nuget package, so you must download and reference it in Wixproj. Let me share with you an example of how I use HeatWave:

```
<Project Sdk="WixToolset.Sdk/4.0.2">
  <PropertyGroup>
    <OutputName>App-Test</OutputName>
    <DefineConstants>
        ProductVersion=$(Version);
    </DefineConstants>
  </PropertyGroup>
  <ItemGroup>
    <HarvestDirectory Include=".\obj\$(Configuration)\publish\YourProject">
      <ComponentGroupName>HarvestedComponents</ComponentGroupName>
      <DirectoryRefId>INSTALLFOLDER</DirectoryRefId>
      <SuppressRootDirectory>true</SuppressRootDirectory>
    </HarvestDirectory>
    <BindPath Include=".\obj\$(Configuration)\publish\YourProject" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="WixToolset.Heat" Version="4.0.0" />
    <PackageReference Include="WixToolset.UI.wixext" Version="4.0.0" />
  </ItemGroup>
 <ItemGroup>
   <ProjectReference Include="..\YourProject\YourProject.csproj" Publish="true" />
 </ItemGroup>
</Project>
```
